File: src/prbm_dfsbfs/WordConversion.java
package prbm_dfsbfs;

public class WordConversion {
	
	public static void main(String[] args) {
//		String begin = "hit";
//		String target = "cog";
//		String[] words = {"hot", "dot", "dog", "lot", "log", "cog"};
		
		String begin = "hit";
		String target = "cog";
		String[] words = {"hot", "dot", "dog", "lot", "log"};
		
		System.out.println(solution(begin, target, words));
	}
	
    public static int solution(String begin, String target, String[] words) {
        int stringLength = begin.length();
        int arrayLength = words.length;
        boolean[] isUsed = new boolean[words.length];
        
        int answer = convert(begin, target, words, isUsed, 0, stringLength, arrayLength);
        
        if(answer<=arrayLength) return answer;
        else return 0;
    }
    
    public static int convert(String before, String target, String[] words, boolean[] isUsed, int depth, int stringLength, int arrayLength) {
		if(isTarget(before, target, stringLength)) return depth;
    	
		if(depth>words.length) return words.length;
    	
    	int min=arrayLength+1;
    	for(int i=0; i<words.length; i++) {
    		if(isUsed[i]==false && isConversible(before, words[i], stringLength)) {
    			isUsed[i]=true;
    			int ret = convert(words[i], target, words, isUsed, depth+1, stringLength, arrayLength);
    			isUsed[i]=false;
    			if(ret<min) min=ret;
    		}
    	}    	   
    	return min;
    }
    
    public static boolean isConversible(String bf, String af, int length) {
    	int j = 0;
    	for(int i=0; i<length; i++) {
    		if(bf.charAt(i)!=af.charAt(i))
    			if(++j>1) return false;
    	}
    	return true;
    }
    
    public static boolean isTarget(String rst, String tgt, int length) {
    	int j = 0;
    	for(int i=0; i<length; i++) {
    		if(rst.charAt(i)!=tgt.charAt(i))
    			if(++j>0) return false;
    	}
    	return true;
    }
    
    

}

File: src/py/pg/prbm_bfsdfs/word_conversion.py
def conversible(bgn, tgt):
    flag = True
    for l, r in zip(bgn, tgt) :
        if not ord(l)-ord(r) : # 같을 때
            continue
        else : # 다를 때
            if flag == False :
                return False # 두 글자 이상 차이남
            flag = False
    return True

def dfs(before, target, words, used, depth):
    # 문제 조건에서 단어수가 max 50이므로
    # 임의의 숫자인 99 사용
    min_cnt = 99
    
    # Case 1&2 중에서 1을 먼저 하는 게 포인트

    # Case1 : 마지막까지 가기 전에 조건을 만족하는 경우    
    if before == target:
        return depth

    # Case2 : 마지막 까지 다 가는 경우
    if depth==len(words):
        return 99
    
    for i, w in enumerate(words):
        if used[i]:
            continue
        if conversible(before, w): 
            used[i] = True
            min_cnt = min(min_cnt, dfs(w, target, words, used, depth+1))
            used[i] = False

    return min_cnt

def solution(begin, target, words):
    used = [False]*len(words)
    answer = dfs(begin, target, words, used, 0)
    return answer if answer!= 99 else 0

print(solution("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]))